llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -1435,6 +1435,51 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
   return nullptr;
 }
 
+Instruction *InstCombinerImpl::simplifyReductionOfShuffle(IntrinsicInst *II) {
+  Intrinsic::ID IID = II->getIntrinsicID();
+  bool CanBeReassociated = (IID != Intrinsic::vector_reduce_fadd &&
+                            IID != Intrinsic::vector_reduce_fmul) ||
+                           II->hasAllowReassoc();
+
+  if (!CanBeReassociated)
+    return nullptr;
+
+  const unsigned ArgIdx = (IID == Intrinsic::vector_reduce_fadd ||
+                           IID == Intrinsic::vector_reduce_fmul)
+                              ? 1
+                              : 0;
+  Value *Arg = II->getArgOperand(ArgIdx);
+  Value *V;
+
```

Comment:
Just a thought but perhaps this will be cleaner is you created something like `simplifyReductionOperand(Value* Op, bool CanReorderLanes)`?  So you'd end up with either:
```
if (auto *NewOp = simplifyReductionOperand(II->getArgOperand(0), true) {
  replaceUse(II->getOperandUse(0), NewOp);
  return nullptr;
}

if (auto *NewOp = simplifyReductionOperand(II->getArgOperand(1), II->hasAllowReassoc()) {
  replaceUse(II->getOperandUse(1), NewOp);
  return nullptr;
}
```

Reply:
That's not quite what I meant.  I was more thinking `simplifyReductionOperand` would take the reduction's operand rather than the reduction itself.  This way the reduction specific parts remain at the call site.

Reply:
Done